src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```diff
@@ -0,0 +1,234 @@
+
+            $type = Type::getType(Type::JSON);
+            $doctrineEnvelope['headers'] = $type->convertToPHPValue($doctrineEnvelope['headers'], $this->driverConnection->getDatabasePlatform());
+
+            $sql = $this->driverConnection->createQueryBuilder()
+                ->update($this->configuration['table_name'])
+                ->set('delivered_at', ':delivered_at')
+                ->where('id = :id');
+            $this->executeQuery($sql, [
+                ':id' => $doctrineEnvelope['id'],
+                ':delivered_at' => (\DateTime::createFromFormat('U.u', microtime(true)))->format('Y-m-d H:i:s.u'),
+            ]);
+
+            $this->driverConnection->commit();
+
+            return $doctrineEnvelope;
+        } catch (\Throwable $e) {
+            $this->driverConnection->rollBack();
+
+            throw $e;
+        }
+    }
+
+    public function ack(string $id): bool
+    {
+        try {
+            return $this->driverConnection->delete($this->configuration['table_name'], ['id' => $id]) > 0;
+        } catch (DBALException $exception) {
+            throw new TransportException($exception->getMessage(), 0, $exception);
+        }
+    }
+
+    public function reject(string $id): bool
+    {
+        try {
+            return $this->driverConnection->delete($this->configuration['table_name'], ['id' => $id]) > 0;
+        } catch (DBALException $exception) {
+            throw new TransportException($exception->getMessage(), 0, $exception);
+        }
+    }
+
+    private function executeQuery(string $sql, array $parameters = [])
+    {
+        $stmt = null;
+        try {
+            $stmt = $this->driverConnection->prepare($sql);
+            $stmt->execute($parameters);
+        } catch (TableNotFoundException $e) {
+            // create table
+            if (!$this->driverConnection->isTransactionActive() && $this->configuration['auto_setup']) {
+                $this->setup();
+            }
+            // statement not prepared ? SQLite throw on exception on prepare if the table does not exist
+            if (null === $stmt) {
+                $stmt = $this->driverConnection->prepare($sql);
+            }
+            $stmt->execute($parameters);
+        }
+
+        return $stmt;
+    }
+
+    public function setup(): void
+    {
+        $synchronizer = new SingleDatabaseSynchronizer($this->driverConnection);
+        $synchronizer->updateSchema($this->getSchema(), true);
+    }
+
+    private function getSchema(): Schema
+    {
+        $schema = new Schema();
+        $table = $schema->createTable($this->configuration['table_name']);
+        $table->addColumn('id', Type::BIGINT)
+            ->setAutoincrement(true)
+            ->setNotnull(true);
+        $table->addColumn('body', Type::TEXT)
+            ->setNotnull(true);
+        $table->addColumn('headers', Type::JSON)
+            ->setNotnull(true);
+        $table->addColumn('queue_name', Type::STRING)
+            ->setLength(32)
```

Comment:
This also seems a bit arbitrary - we could have some people creating really long queue names

Reply:
Yes. I can use `255` but that's also arbitrary ... or make it configurable with a default value of 255 ? :wink:

Reply:
In doctrine/dbal default is 255 : https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Platforms/AbstractPlatform.php#L595

Reply:
yea, that's why I was thinking 255. I think generally people expect that 255 is the make for a normal string field in a database, so it's probably not surprising that this would be the length. 32 is short enough that someone might try something longer and it'll truncate. 

It's *sort* of already configurable... because they could just create the table with a shorter/longer length manually.